### PR TITLE
fix(multi-tenancy): Make proposals for delete namespace request

### PR DESCRIPTION
**Pull request by ahsanbarkati**
This PR fixes delete namespace operation. The delete needs to be proposed to
all the groups. And within the group, we need a raft proposal for delete operation.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7417)

### CODE REVIEW

File: protos/pb.proto [hunk 1]
```diff
@@ -319,6 +319,7 @@ message Proposal {
 	uint64 index           		= 10; // Used to store Raft index, in raft.Ready.
 	uint64 expected_checksum 	= 11; // Block an operation until membership reaches this checksum.
 	RestoreRequest restore 		= 12;
+	DeleteNsRequest delete 		= 13; // Used to delete namespace.
```
Comment (NamanJain8) [inline protos/pb.proto:322]: Maybe rename delete --> deleteNs

File: worker/multi_tenancy.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
+// +build !oss
+
+/*
+ * Copyright 2018 Dgraph Labs, Inc. and Contributors
```
Comment (NamanJain8) [inline worker/multi_tenancy.go:4]: 2021

File: worker/multi_tenancy.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import (
+	"context"
+	"time"
+
+	"github.com/dgraph-io/dgraph/conn"
+	"github.com/dgraph-io/dgraph/protos/pb"
+	"github.com/pkg/errors"
+)
+
+func (w *grpcWorker) DeleteNamespace(ctx context.Context,
```
Comment (NamanJain8) [inline worker/multi_tenancy.go:24]: Add declaration of all these functions in some file with `+build oss`. Those functions may return not supported error.

**Review by NamanJain8 — APPROVED** [review]

File: worker/multi_tenancy.go (hunk 3 above)
Comment (ahsanbarkati) [reply worker/multi_tenancy.go:27]: Thanks. Done!

**Review by jarifibrahim — APPROVED** [review]

File: worker/multi_tenancy.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"context"
+
+	"github.com/dgraph-io/dgraph/protos/pb"
+)
+
+func (w *grpcWorker) DeleteNamespace(ctx context.Context,
+	req *pb.DeleteNsRequest) (*pb.Status, error) {
+	return nil, nil
+}
```
Comment (NamanJain8) [inline worker/multi_tenancy.go:30]: Shouldn't we return an error here, like `ErrNotSupported`? (see Backup)

**Review by NamanJain8 — APPROVED** [review]

File: worker/multi_tenancy.go (hunk 4 above)
Comment (ahsanbarkati) [reply worker/multi_tenancy.go:30]: Done

File: worker/multi_tenancy.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+func ProcessDeleteNsRequest(ctx context.Context, ns uint64) error {
+	// Update the membership state to get the latest mapping of groups to predicates.
+	if err := UpdateMembershipState(ctx); err != nil {
+		return errors.Wrapf(err, "Failed to update membership state while deleting namesapce")
+	}
+
+	state := GetMembershipState()
+
+	errCh := make(chan error, len(state.Groups))
```
[the remark this answers is not in the document]
Comment (ahsanbarkati) [reply worker/multi_tenancy.go]: This is nice!

File: worker/multi_tenancy.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		}
+	}
+	return nil
+}
+
+// TODO - Currently there is no way by which we can prevent partial banning. It is possible
+// that if a group is behind a network partition then we might not be able to propose to it.
+func tryDeleteProposal(ctx context.Context, req *pb.DeleteNsRequest) error {
+	var err error
+	for i := 0; i < 10; i++ {
```
[the remark this answers is not in the document]
Comment (ahsanbarkati) [reply worker/multi_tenancy.go]: Done!

File: worker/multi_tenancy.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	for gid, _ := range state.Groups {
+		req := &pb.DeleteNsRequest{Namespace: ns, GroupId: gid}
+		go func(req *pb.DeleteNsRequest) {
+			errCh <- tryDeleteProposal(ctx, req)
+		}(req)
+	}
+
+	for range state.Groups {
+		if err := <-errCh; err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (ahsanbarkati) [reply worker/multi_tenancy.go]: Done!

**Pull request merged**